File: starter_kits/C#/Core/Position.cs
using System;

namespace Halite3.Core
{
    public class Position
    {
        public int X { get; private set; }
        public int Y { get; private set; }

        public Position(int x, int y)
        {
            this.X = x;
            this.Y = y;
        }

        public Position DirectionalOffset(Direction d)
        {
            int dx;
            int dy;

            switch (d)
            {
                case Direction.NORTH:
                    dx = 0;
                    dy = -1;
                    break;
                case Direction.SOUTH:
                    dx = 0;
                    dy = 1;
                    break;
                case Direction.EAST:
                    dx = 1;
                    dy = 0;
                    break;
                case Direction.WEST:
                    dx = -1;
                    dy = 0;
                    break;
                case Direction.STILL:
                    dx = 0;
                    dy = 0;
                    break;
                default:
                    throw new ArgumentException("Unknown direction " + d);
            }

            return new Position(X + dx, Y + dy);
        }

        public override bool Equals(object other)
        {
            if (this == other) return true;
            if (!(other is Position)) return false;

            var position = other as Position;

            if (X != position.X) return false;
            return Y == position.Y;
        }

        public override int GetHashCode()
        {
            int result = X;
            result = 31 * result + Y;
            return result;
        }
    }
}

File: starter_kits/Swift/Halite-III/Halite-III/hlt/Position.swift
//
//  Position.swift
//  Halite-III
//
//  Created by Chris Downie on 10/24/18.
//  Copyright © 2018 Chris Downie. All rights reserved.
//

import Foundation

struct Position {
    let x: Int
    let y: Int
    
    /// Returns a new position based on moving one unit in the given direction from the given position.
    ///
    /// - Parameter direction: The direction to move this position
    /// - Returns: The position moved one unit from the given direction
    func directionalOffset(_ direction: Direction) -> Position {
        return self + direction
    }
    
    /// Returns a list of all positions around the given position in each cardinal direction.
    ///
    /// - Returns: A list of all positions around the given position in each cardinal direction.
    func getSurroundingCardinals() -> [Position] {
        return Direction.allCardinals.map { direction -> Position in
            return self + direction
        }
    }
}

// MARK: - Equatable
extension Position: Equatable {}

// MARK: - Some numeric-style methods
extension Position {
    static func +(_ position: Position, _ direction: Direction) -> Position {
        var x = position.x, y = position.y
        switch direction {
        case .north:
            y -= 1
        case .south:
            y += 1
        case .east:
            x += 1
        case .west:
            x -= 1
        case .still:
            break;
        }
        return Position(x: x, y: y)
    }
    
    static func +(_ direction: Direction, _ position: Position) -> Position {
        return position + direction
    }
    
    static func +=(_ position: inout Position, _ direction: Direction) {
        position = position + direction
    }
    
    static func -(_ position: Position, _ direction: Direction) -> Position {
        return position + Direction.invert(direction)
    }
    
    static func -=(_ position: inout Position, _ direction: Direction) {
        position = position - direction
    }
}

// MARK: - Convenience Methods
extension Position {
    var surroundingCardinals: [Position] {
        return getSurroundingCardinals()
    }
}
